nearform/mercurius-federation-info-graphiql-plugin
# feat: Update tree view layout according to design

This PR closes #83 

It updates the service panel using the MUI components.

For the tree view, it uses the MUI's `TreeView`. 

<img width="1436" alt="Screenshot 2022-12-07 at 14 38 07" src="https://user-images.githubusercontent.com/1233582/206194034-72bfd5ee-74dc-49f2-a3c2-4411fe4c6e04.png">

<img width="1435" alt="Screenshot 2022-12-07 at 14 38 16" src="https://user-images.githubusercontent.com/1233582/206194062-b8e72c55-f82f-4bd7-80b8-3316fdbd93b6.png">

<img width="1437" alt="Screenshot 2022-12-07 at 14 38 51" src="https://user-images.githubusercontent.com/1233582/206196135-2648282b-0292-469b-bf7d-905215635fb1.png">

The plugin now listens to GraphiQL theme updates (changing from `light` to `dark` and vice-versa) and updates the MUI theme. The hook provided by GraphiQL is bugged, so if you change the theme from the settings, you have to close and reopen the plugin's tab to get the updated value (GraphiQL [issue](https://github.com/graphql/graphiql/issues/2956)).

I've also introduced a `PluginStateContext` that stores the open and closed services and their trees in the `LocalStorage`; in this way, if the user moves to another GraphiQL tab or closes and reopens the browser, the app restores the state. 

About the items "decorators" like `@extends`, `@external` and `@key(...)`, I converted them into MUI `<Chip>` components. For the `@key(...)` one, I discussed with @matijagaspar, to:
- move them from the type name to the actual field.
- use a more compacted representation using `Chip` with a "key" icon.

This is the actual implementation in `master`:
![Screenshot_2022-12-05_at_12_27_37](https://user-images.githubusercontent.com/1233582/206225382-7e0136b5-0e53-43c7-a6fe-36cbe7a6721f.png)

This is the one proposed by this PR:

<img width="431" alt="Graphiql_Federation_Info_Demo-2" src="https://user-images.githubusercontent.com/1233582/206226401-c5f50b89-7fe3-4451-9b33-8affb75cbcf9.png">


The PR also:
- removes all the SCSS files used by the service components; I've replaced all `div` elements with MUI's `<Box>` and used the `sx` prop to add custom CSS styles.
- adds the total number of services next to the panel's title.
- for each service, adds the total number of items in the toggle button.

# How to test it

To test this PR:
- install the dependencies with `npm ci` 
- execute the command `npm run example` 
- open your browser to http://localhost:3001/graphiql
- then click on the plugin tab
<img width="706" alt="GraphiQL" src="https://user-images.githubusercontent.com/1233582/206235091-03d5ee8b-4c5a-4cf0-b217-465fa0384702.png">

Please note that this solution does not require the `mercurius-federation-info` project to run because the plugin example builds the source code and runs a mercurius instance for you.
